File: backend/app/schemas/project.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.models.project import ProjectMemberRole, ProjectVideoMode


class ProjectBase(BaseModel):
    name: str = Field(min_length=1, max_length=100, description="项目名称")
    intro: str = Field(default="", max_length=2000, description="项目简介")
    project_type: str = Field(default="novel_to_video", description="项目类型")
    content_type: str = Field(default="novel", description="内容来源类型")
    art_style: str = Field(default="3D_chinese_traditional", description="艺术风格")
    director_manual: str = Field(default="", description="导演风格")
    video_ratio: str = Field(default="9:16", description="视频比例")
    text_model: str = Field(default="", description="文本模型")
    image_model: str = Field(default="", description="生图模型")
    video_model: str = Field(default="", description="视频模型")
    tts_model: str = Field(default="", description="语音模型")
    image_quality: str = Field(default="standard", description="图片质量")
    mode: ProjectVideoMode = Field(default=ProjectVideoMode.TEXT, description="视频生成模式")


class ProjectCreate(ProjectBase):
    pass


class ProjectUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=100)
    intro: str | None = None
    project_type: str | None = None
    content_type: str | None = None
    art_style: str | None = None
    director_manual: str | None = None
    video_ratio: str | None = None
    text_model: str | None = None
    image_model: str | None = None
    video_model: str | None = None
    tts_model: str | None = None
    image_quality: str | None = None
    mode: ProjectVideoMode | None = None


class ProjectRead(ProjectBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    public_id: str
    owner_id: str
    my_role: ProjectMemberRole | None = None
    sort_order: int
    created_at: datetime
    updated_at: datetime
    disabled_at: datetime | None


class ProjectMemberRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    project_id: int
    user_public_id: str
    user_email: str | None
    role: ProjectMemberRole
    joined_at: datetime


class ProjectMemberInvite(BaseModel):
    user_public_id: str = Field(min_length=1, max_length=36, description="被邀请用户的公开标识")
    role: ProjectMemberRole = Field(default=ProjectMemberRole.VIEWER, description="邀请加入后的项目角色")


class VisualStyleFileRead(BaseModel):
    """视觉风格 Markdown 文件读取结果。"""

    path: str = Field(description="相对于风格目录的 Markdown 文件路径。")
    content: str = Field(description="Markdown 文件内容。")
    size_bytes: int = Field(ge=0, description="文件大小，单位字节。")


class VisualStyleFileWrite(BaseModel):
    """视觉风格 Markdown 文件写入请求。"""

    path: str = Field(default="", max_length=255, description="相对于风格目录的 Markdown 文件路径。")
    content: str = Field(default="", description="Markdown 文件内容。")

    @field_validator("path")
    @classmethod
    def validate_markdown_path(cls, value: str) -> str:
        if value and not value.lower().endswith(".md"):
            raise ValueError("视觉风格文件必须是 .md 文件")
        return value


class VisualStyleImageRead(BaseModel):
    """视觉风格图片读取结果。"""

    filename: str = Field(description="图片文件名。")
    path: str = Field(description="相对于风格目录的图片路径。")
    url: str = Field(description="可直接用于 CSS url() 或 HTML src 的图片访问 URL。")
    size_bytes: int = Field(ge=0, description="文件大小，单位字节。")


class VisualStyleImageWrite(BaseModel):
    """视觉风格图片写入请求。"""

    filename: str = Field(default="", max_length=120, description="图片文件名。")
    data: str = Field(description="图片 base64 内容，可包含 data URI 前缀。")


class VisualStyleRead(BaseModel):
    """视觉风格完整读取结果。"""

    style_path: str = Field(description="视觉风格目录名。")
    name: str = Field(default="", description="视觉风格展示名称。")
    files: list[VisualStyleFileRead] = Field(default_factory=list, description="Markdown 文件列表。")
    images: list[VisualStyleImageRead] = Field(default_factory=list, description="图片文件列表。")


class VisualStyleCreate(BaseModel):
    """创建视觉风格请求。"""

    style_path: str = Field(min_length=1, max_length=120, description="视觉风格目录名。")
    name: str = Field(default="", max_length=120, description="视觉风格展示名称。")
    files: list[VisualStyleFileWrite] = Field(default_factory=list, description="需要写入的 Markdown 文件。")
    images: list[VisualStyleImageWrite] = Field(default_factory=list, description="需要写入的图片。")


class VisualStyleUpdate(BaseModel):
    """更新视觉风格请求。"""

    name: str | None = Field(default=None, max_length=120, description="视觉风格展示名称。")
    files: list[VisualStyleFileWrite] | None = Field(default=None, description="需要新增或覆盖的 Markdown 文件。")
    images: list[VisualStyleImageWrite] | None = Field(default=None, description="需要新增或覆盖的图片。")
    replace_images: bool = Field(default=False, description="是否先清空 images 目录下已有图片。")


class DirectorManualFileRead(BaseModel):
    """导演手册 Markdown 文件读取结果。"""

    path: str = Field(description="相对于导演手册目录的 Markdown 文件路径。")
    content: str = Field(description="Markdown 文件内容。")
    size_bytes: int = Field(ge=0, description="文件大小，单位字节。")


class DirectorManualFileWrite(BaseModel):
    """导演手册 Markdown 文件写入请求。"""

    path: str = Field(default="", max_length=255, description="相对于导演手册目录的 Markdown 文件路径。")
    content: str = Field(default="", description="Markdown 文件内容。")

    @field_validator("path")
    @classmethod
    def validate_markdown_path(cls, value: str) -> str:
        if value and not value.lower().endswith(".md"):
            raise ValueError("导演手册文件必须是 .md 文件")
        return value


class DirectorManualImageRead(BaseModel):
    """导演手册图片读取结果。"""

    filename: str = Field(description="图片文件名。")
    path: str = Field(description="相对于导演手册目录的图片路径。")
    url: str = Field(description="可直接用于 CSS url() 或 HTML src 的图片访问 URL。")
    size_bytes: int = Field(ge=0, description="文件大小，单位字节。")


class DirectorManualImageWrite(BaseModel):
    """导演手册图片写入请求。"""

    filename: str = Field(default="", max_length=120, description="图片文件名。")
    data: str = Field(description="图片 base64 内容，可包含 data URI 前缀。")


class DirectorManualRead(BaseModel):
    """导演手册完整读取结果。"""

    manual_path: str = Field(description="导演手册目录名。")
    name: str = Field(default="", description="导演手册展示名称。")
    files: list[DirectorManualFileRead] = Field(default_factory=list, description="Markdown 文件列表。")
    images: list[DirectorManualImageRead] = Field(default_factory=list, description="图片文件列表。")


class DirectorManualCreate(BaseModel):
    """创建导演手册请求。"""

    manual_path: str = Field(min_length=1, max_length=120, description="导演手册目录名。")
    name: str = Field(default="", max_length=120, description="导演手册展示名称。")
    files: list[DirectorManualFileWrite] = Field(default_factory=list, description="需要写入的 Markdown 文件。")
    images: list[DirectorManualImageWrite] = Field(default_factory=list, description="需要写入的图片。")

File: frontend/src/api/project.ts
import request from '@/request'

export type ProjectVideoMode =
  | 'text'
  | 'singleImage'
  | 'multiReference'
  | 'startEndRequired'
  | 'endFrameOptional'
  | 'startFrameOptional'

export type ProjectMemberRole = 'owner' | 'admin' | 'manager' | 'editor' | 'viewer'

export interface ProjectRecord {
  id: number
  public_id: string
  owner_id: string
  name: string
  intro: string
  project_type: string
  content_type: string
  art_style: string
  director_manual: string
  video_ratio: string
  text_model: string
  image_model: string
  video_model: string
  tts_model: string
  image_quality: string
  mode: ProjectVideoMode
  my_role?: ProjectMemberRole | null
  sort_order: number
  created_at: string
  updated_at: string
  disabled_at?: string | null
}

export interface ProjectPayload {
  name: string
  intro: string
  project_type: string
  content_type: string
  art_style: string
  director_manual: string
  video_ratio: string
  text_model: string
  image_model: string
  video_model: string
  tts_model: string
  image_quality: string
  mode: ProjectVideoMode
}

export interface ProjectMemberRecord {
  id: number
  project_id: number
  user_public_id: string
  user_email: string | null
  role: ProjectMemberRole
  joined_at: string
}

export interface ProjectMemberInvitePayload {
  user_public_id: string
  role: ProjectMemberRole
}

export interface UserRecord {
  public_id: string
  username: string
  nickname?: string | null
  email?: string | null
  avatar_url?: string | null
  sort_order: number
  is_superuser: boolean
  disabled_at?: string | null
  last_login_at?: string | null
  created_at: string
  updated_at: string
}

export interface VisualStyleFileRecord {
  path: string
  content: string
  size_bytes: number
}

export interface VisualStyleImageRecord {
  filename: string
  path: string
  url: string
  size_bytes: number
}

export interface VisualStyleRecord {
  style_path: string
  name: string
  files: VisualStyleFileRecord[]
  images: VisualStyleImageRecord[]
}

export interface VisualStyleFilePayload {
  path: string
  content: string
}

export interface VisualStyleImagePayload {
  filename: string
  data: string
}

export interface VisualStyleCreatePayload {
  style_path: string
  name: string
  files: VisualStyleFilePayload[]
  images: VisualStyleImagePayload[]
}


export interface DirectorManualFileRecord {
  path: string
  content: string
  size_bytes: number
}

export interface DirectorManualImageRecord {
  filename: string
  path: string
  url: string
  size_bytes: number
}

export interface DirectorManualRecord {
  manual_path: string
  name: string
  files: DirectorManualFileRecord[]
  images: DirectorManualImageRecord[]
}

export interface DirectorManualFilePayload {
  path: string
  content: string
}

export interface DirectorManualImagePayload {
  filename: string
  data: string
}

export interface DirectorManualCreatePayload {
  manual_path: string
  name: string
  files: DirectorManualFilePayload[]
  images: DirectorManualImagePayload[]
}


// export 表示把当前函数暴露给外界，允许其他模块通过 import {函数名} from "模块名"进行调用
export const listProjectsApi = () => request.get<ProjectRecord[]>('/projects/')

export const listVisualStylesApi = () => request.get<VisualStyleRecord[]>('/projects/visual-styles')

export const createVisualStyleApi = (data: VisualStyleCreatePayload) => (
  request.post<VisualStyleRecord>('/projects/visual-styles', data)
)

export const getVisualStyleApi = (stylePath: string) => (
  request.get<VisualStyleRecord>(`/projects/visual-styles/${encodeUrlSegment(stylePath)}`)
)

export const writeVisualStyleFileApi = (
  stylePath: string,
  filePath: string,
  data: VisualStyleFilePayload,
) => (
  request.put<VisualStyleFileRecord>(
    `/projects/visual-styles/${encodeUrlSegment(stylePath)}/files/${encodeUrlPath(filePath)}`,
    data,
  )
)

export const deleteVisualStyleFileApi = (stylePath: string, filePath: string) => (
  request.delete(`/projects/visual-styles/${encodeUrlSegment(stylePath)}/files/${encodeUrlPath(filePath)}`)
)

export const writeVisualStyleImageApi = (
  stylePath: string,
  filename: string,
  data: VisualStyleImagePayload,
) => (
  request.put<VisualStyleImageRecord>(
    `/projects/visual-styles/${encodeUrlSegment(stylePath)}/images/${encodeUrlSegment(filename)}`,
    data,
  )
)

export const deleteVisualStyleImageApi = (stylePath: string, filename: string) => (
  request.delete(`/projects/visual-styles/${encodeUrlSegment(stylePath)}/images/${encodeUrlSegment(filename)}`)
)

export const listDirectorManualsApi = () => request.get<DirectorManualRecord[]>('/projects/director-manuals')

export const createDirectorManualApi = (data: DirectorManualCreatePayload) => (
  request.post<DirectorManualRecord>('/projects/director-manuals', data)
)

export const writeDirectorManualFileApi = (
  manualPath: string,
  filePath: string,
  data: DirectorManualFilePayload,
) => (
  request.put<DirectorManualFileRecord>(
    `/projects/director-manuals/${encodeUrlSegment(manualPath)}/files/${encodeUrlPath(filePath)}`,
    data,
  )
)

export const deleteDirectorManualFileApi = (manualPath: string, filePath: string) => (
  request.delete(`/projects/director-manuals/${encodeUrlSegment(manualPath)}/files/${encodeUrlPath(filePath)}`)
)

export const writeDirectorManualImageApi = (
  manualPath: string,
  filename: string,
  data: DirectorManualImagePayload,
) => (
  request.put<DirectorManualImageRecord>(
    `/projects/director-manuals/${encodeUrlSegment(manualPath)}/images/${encodeUrlSegment(filename)}`,
    data,
  )
)

export const deleteDirectorManualImageApi = (manualPath: string, filename: string) => (
  request.delete(`/projects/director-manuals/${encodeUrlSegment(manualPath)}/images/${encodeUrlSegment(filename)}`)
)

export const searchProjectsByNameApi = (name: string) => (
  request.get<ProjectRecord[]>('/projects/search/by-name', { params: { name } })
)

export const createProjectApi = (data: ProjectPayload) => request.post<ProjectRecord>('/projects/', data)

export const updateProjectApi = (publicId: string, data: Partial<ProjectPayload>) => (
  request.put<ProjectRecord>(`/projects/${publicId}`, data)
)

export const disableProjectApi = (publicId: string) => request.patch<ProjectRecord>(`/projects/${publicId}/disable`)

export const enableProjectApi = (publicId: string) => request.patch<ProjectRecord>(`/projects/${publicId}/enable`)

export const deleteProjectApi = (publicId: string) => request.delete(`/projects/${publicId}`)

export const listProjectMembersApi = (publicId: string) => (
  request.get<ProjectMemberRecord[]>(`/projects/${publicId}/members`)
)

export const searchProjectMemberCandidatesApi = (publicId: string, keyword: string) => (
  request.get<UserRecord[]>(`/projects/${publicId}/members/candidates`, { params: { keyword } })
)

export const inviteProjectMemberApi = (publicId: string, data: ProjectMemberInvitePayload) => (
  request.post<ProjectMemberRecord>(`/projects/${publicId}/members/invitations`, data)
)

export const updateProjectMemberRoleApi = (
  publicId: string,
  userPublicId: string,
  role: ProjectMemberRole,
) => (
  request.patch<ProjectMemberRecord>(`/projects/${publicId}/members/${userPublicId}`, null, { params: { role } })
)

export const removeProjectMemberApi = (publicId: string, userPublicId: string) => (
  request.delete(`/projects/${publicId}/members/${userPublicId}`)
)

const encodeUrlSegment = (value: string) => encodeURIComponent(value.trim())

const encodeUrlPath = (value: string) => (
  value
    .trim()
    .replace(/\\/g, '/')
    .split('/')
    .map(encodeURIComponent)
    .join('/')
)
